### PR TITLE
Filter by insertions

### LINKED ISSUE #318: Filter by insertions
https://github.com/GenSpectrum/LAPIS-SILO/pull/166 made it possible to filter sequences by insertions. Expose this via LAPIS.

How do we implement this in LAPIS 2? The old LAPIS offers two fields `nucInsertions` and `aaInsertions`, but SILO has columns of type `insertion` that need to be configured.

### PR DESCRIPTION
**Pull request by JonasKellerer**
Resolves #318

### CODE REVIEW

**Comment by JonasKellerer** [comment]
This PR waits on SILO, so that it can cope with multiple nucleotideInsertion filters.

File: lapis2/src/main/kotlin/org/genspectrum/lapis/request/NucleotideInsertion.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                insertions,
+                matchGroups["segment"]?.value,
+            )
+        }
+    }
+}
+
+private val NUCLEOTIDE_INSERTION_REGEX =
+    Regex(
+        """^ins_((?<segment>[a-zA-Z0-9_-]+)(?=:):)?(?<position>\d+):(?<insertions>[a-zA-Z0-9?_-]+)?$""",
```
Comment (fengelniederhammer) [inline lapis2/src/main/kotlin/org/genspectrum/lapis/request/NucleotideInsertion.kt:39]: Why is `_` and `-` allowed in the insertion? And should we maybe also allow the SILO syntax `.*`?

File: lapis2/src/main/kotlin/org/genspectrum/lapis/OpenApiDocs.kt [hunk 2]
```diff
@@ -242,6 +254,35 @@ private fun aminoAcidMutations() =
...
+private fun nucleotideInsertions() =
+    Schema<List<NucleotideInsertion>>()
+        .type("array")
+        .items(
+            Schema<String>()
+                .type("string")
+                .example("ins_123:ATT")
+                .description(
+                    """
+                    |A nucleotide insertion in the format "ins_\<sequenceName\>?:\<position\>?:\<insertion\>?".
```
Comment (fengelniederhammer) [inline lapis2/src/main/kotlin/org/genspectrum/lapis/OpenApiDocs.kt:266]: Isn't this slightly more correct?
```suggestion
                    |A nucleotide insertion in the format "ins_(\<sequenceName\>:)?\<position\>:\<insertion\>".  
```

File: lapis2/src/main/kotlin/org/genspectrum/lapis/OpenApiDocs.kt [hunk 3]
```diff
@@ -242,6 +254,35 @@ private fun aminoAcidMutations() =
...
+private fun aminoAcidInsertions() =
+    Schema<List<AminoAcidInsertion>>()
+        .type("array")
+        .items(
+            Schema<String>()
+                .type("string")
+                .example("ins_ORF1:123:ATT")
+                .description(
+                    """
+                    |A amino acid insertion in the format "ins_\<gene\>?:\<position\>?:\<insertion\>?".
```
Comment (fengelniederhammer) [inline lapis2/src/main/kotlin/org/genspectrum/lapis/OpenApiDocs.kt:281]: Same here: 
```suggestion
                    |A amino acid insertion in the format "ins_\<gene\>:\<position\>:\<insertion\>".  
```

**Review by fengelniederhammer — APPROVED** [review]
Looks good.

File: lapis2/src/test/kotlin/org/genspectrum/lapis/controller/LapisControllerTest.kt [hunk 4]
```diff
@@ -275,11 +342,27 @@ class LapisControllerTest(@Autowired val mockMvc: MockMvc) {
...
         )
+
+        @JvmStatic
+        fun getEndpointsWithInsertionFilter() = allEndpoints()
+
+        @JvmStatic
+        fun getEndpointsWithNucleotideMutationFilter() = allEndpoints()
+
+        @JvmStatic
+        fun getEndpointsWithAminoAcidMutationFilter() = allEndpoints()
```
[the remark this answers is not in the document]
Comment (JonasKellerer) [reply lapis2/src/test/kotlin/org/genspectrum/lapis/controller/LapisControllerTest.kt:365]: This is done partially and naivly in #364 for the csv/tsv endpoints. I introduce a new ticket for these tests, see #378 #379 and #380

**Pull request merged**